### PR TITLE
chore(publish) abstract environment variables and introduce feature flag

### PR DESCRIPTION
**Pull request by dduportal**
Related to https://github.com/jenkins-infra/helpdesk/issues/2649#issuecomment-2142644635

This PR is a follow-up of #784 and is were review should happen (cc @daniel-beck ) as we use the PR here to test changed in a dummy freestyle job in trusted.ci.

The goal is the following:

> While working on https://github.com/jenkins-infra/update-center2/pull/777 for the [Update Center migration to a mirror system](https://github.com/jenkins-infra/helpdesk/issues/2649), we realized the following problems with the current `publish.sh` script:
> 
> - It's hard to test a change until it's in production => we would want to enable/disable some steps to only focus testing on others in order to have enough confidence to merge PRs without breaking the production (neither requiring a fully fledged staging environment)
> - There are parameters written in the script itself which requires SREs to open an (untestable for today) PR and merge it as soon as possible when rotating expired credentials/changing configurations of bucket, etc.
> 
> 
> This PR is a proposal to solve these issues with:
> 
> - Feature flagging for the top-levels stages of this script, through en environment variable which has a default value (same behavior as today) and can be overridden by the caller
> - Definition of the Update Center sync. tasks through an environment variable (same as above: default value which can be overridden) and a convention that each "sync" task will load their own environment file to get their parameters and credentials
> 

See https://github.com/jenkins-infra/charts-secrets/pull/30 for the credential generation

Please note this PR also adds the generation of the `www-redirections/` folder which is aimed to HTTPD (to provide the redirection logic without providing files which are expected to be mirrored) from https://github.com/jenkins-infra/update-center2/pull/777

The following diagram represents the different "top level" stages and the data passed between them.
It's key to understand what kind of testing we are doing for the "Sync UC" stage by mocking the `./www2/` directory.

```                                                                                              
                                                                                              
    ┌────────────────────────────┐                         ┌─────────────────────────────┐    
    │                            │                         │                             │    
    │                            ├───────./download/───────►                             │    
    │                            │                         │                             │    
    │                            │                         │                             │    
    │         Generate Site      │                         │       Sync Plugins          │    
    │                            │                         │                             │    
    │                            │                         │                             │    
    │                            ├────./www2/──────┬──────►│                             │    
    │                            │                 │       │                             │    
    └────────────────────────────┘                 │       └─────────────────────────────┘    
                                                   │                                          
                                                   │                                          
                                                   │                                          
                                                   │                                          
                                                   │                                          
                                                   │                                          
                                                   │                                          
                                                   │                                          
                             ┌─────────────────────▼──────────────┐                           
                             │                                    │                           
                             │                                    │                           
                             │                                    │                           
                             │                                    │                           
                             │                                    │                           
                             │                                    │                           
                             │            Sync Update Center      │                           
                             │                                    │                           
                             │                                    │                           
                             │                                    │                           
                             │                                    │                           
                             │                                    │                           
                             └────────────────────────────────────┘                           
```

### CODE REVIEW

**Comment by dduportal** [comment]
This PR is ready to review and to be merged:

TEST:
- Credential setup:
  - [x] Generated the ZIP file from https://github.com/jenkins-infra/charts-secrets/blob/main/config/trusted.ci.jenkins.io/get-uc-sync-zip-credential.sh
  - [x] Updated the `.env-rsync-updates.jenkins.io` file (changing the source dir and the destination dir), recrerated the ZIP and added it a scoped **folder**-level credential named `update-center-ucsync-tasks-pr-787` in https://trusted.ci.jenkins.io:1443/job/update_center_test_lemeurherve_helpdesk2649/
  - [x] Set up the trusted.ci's [`update_center_test_lemeurherve_helpdesk2649`](https://trusted.ci.jenkins.io:1443/job/update_center_test_lemeurherve_helpdesk2649/job/update_center_PR_784/) Test freestyle job to use this new credential (secret ZIP file, using the variable `UPDATE_CENTER_FILESHARES_ENV_FILES`)
- Set up the Test job [`update_center_test_lemeurherve_helpdesk2649`](https://trusted.ci.jenkins.io:1443/job/update_center_test_lemeurherve_helpdesk2649/job/update_center_PR_784/) with:
  - Retrieving a fresh set of "real life" data into the workspace's local `uc-www` (to test WWW2 overriding)
  - Do not run the generate and sync-plugins stages: only sync-uc from the `uc-www` (to test RUN_STAGES overriding and our env credentials)

  ```bash
  # This WWW2 dir also needs to be set in the .env-rsync-updates.jenkins.io (soon to be deleted so no need to be smart)
  export WWW2_DIR="${WORKSPACE}/uc-www/"
  export RUN_STAGES='sync-uc'
  # Retrieve fresh "real" UC www data from the remote pkg.origin.jenkins.io
  rsync -avz --stats --delete mirrorbrain@updates.jenkins.io:/var/www/updates.jenkins.io/ "${WWW2_DIR}"/
  # Ensure the remote "mock" directory is available on the pkg VM (and smoke test the env)
  {
    . "${UPDATE_CENTER_FILESHARES_ENV_FILES}/.env-rsync-updates.jenkins.io"
    ssh ${RSYNC_USER}@${RSYNC_HOST} mkdir -p "${RSYNC_REMOTE_DIR}"
  }
  # Execute
  CERTIFICATE_MINIMUM_VALID_DAYS=30 RECENT_RELEASES_MAX_AGE_HOURS=6 PATH=${JDK8_HOME}/bin:${MVN_HOME}/bin:$PATH bash -x site/publish.sh
  ```
  - Ran successfully the job with multiple configurations (to test `SYNC_UC_TASKS`) and with the default (to check production-like default setup)

PRODUCTION (so this PR can be merged without breaking it):

- [x] Added a new top-level credential named `update-center-publish-env`
  - [x] Set up the trusted.ci's `update_center` freestyle job to use this new credential (secret ZIP file, using the variable `UPDATE_CENTER_FILESHARES_ENV_FILES`)

CLEANUP

Once merged and running successfully, we'll have to clean up:
- Removing the folder https://trusted.ci.jenkins.io:1443/job/update_center_test_lemeurherve_helpdesk2649 (and its scoped credentials, and its workspace)
- Remove the `ddu-test` temp directories on both pkg VM and trusted agent
- Remove the following (unused anymore) credentials (both update_center credential binding AND credential):
  - `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` (both are embeded in the new ZIP credential file in `.env-s3sync-westeurope`)
  - `JENKINS_INFRA_FILESHARE_*` Azure service principal (embeded in the new ZIP credential file in `.env-azsync-content`)

File: site/publish.sh [hunk 1]
```diff
@@ -1,147 +1,179 @@
...
+                --delete-destination=true \
+                "${fileshare_sync_source_abs}" "${fileShareUrl}"
+            ;;
+
+        s3sync*)
+            # Required variables that should now be set from the .env file
+            : "${BUCKET_NAME?}" "${BUCKET_ENDPOINT_URL?}" "${AWS_ACCESS_KEY_ID?}" "${AWS_SECRET_ACCESS_KEY?}" "${AWS_DEFAULT_REGION?}"
+
+            # Sync 'www-content' (without symlinks) to the bucket,
+            # excluding 'updates/' folderas it is populated by https://github.com/jenkins-infra/crawler/blob/master/Jenkinsfile
```
Comment (smerle33) [inline site/publish.sh:111]:
```suggestion
            # excluding 'updates/' folders it is populated by https://github.com/jenkins-infra/crawler/blob/master/Jenkinsfile
```

Comment (dduportal) [reply site/publish.sh:111]:
```suggestion
            # excluding 'updates/' folder as it is populated by https://github.com/jenkins-infra/crawler/blob/master/Jenkinsfile
```

File: site/publish.sh [hunk 2]
```diff
@@ -1,147 +1,179 @@
...
+    rsync -rlptDvz --chown=mirrorbrain:www-data --size-only "${download_dir}"/plugins/ "${RSYNC_USER}@${UPDATES_SITE}":/srv/releases/jenkins/plugins
+
+    # Invoke a minimal mirrorsync to mirrorbits which will use the 'recent-releases.json' file as input
+    ssh "${RSYNC_USER}@${UPDATES_SITE}" "cat > /tmp/update-center2-rerecent-releases.json" < "${www2_dir}"/experimental/recent-releases.json
+    ssh "${RSYNC_USER}@${UPDATES_SITE}" "/srv/releases/sync-recent-releases.sh /tmp/update-center2-rerecent-releases.json"
+fi
+
+if [[ "${run_stages[*]}" =~ 'sync-uc' ]]
+then
+    # Ensure credentials is defined
```
Comment (timja) [inline site/publish.sh:47]:
```suggestion
    # Ensure credentials are defined
```

File: site/publish.sh [hunk 3]
```diff
@@ -1,147 +1,179 @@
...
+    : "${UPDATE_CENTER_FILESHARES_ENV_FILES?}"
+
+    sync_uc_tasks=()
+    IFS='|' read -r -a sync_uc_tasks <<< "${SYNC_UC_TASKS}"
+
+    # parallel added within the permanent trusted agent here:
+    # https://github.com/jenkins-infra/jenkins-infra/blob/production/dist/profile/manifests/buildagent.pp
```
Comment (timja) [inline site/publish.sh:54]: not required imo
```suggestion
```

**Review by timja — APPROVED** [review]
LGTM

**Review by smerle33 — APPROVED** [review]
🚀

**Comment by dduportal** [comment]
Given this change blocks the Jenkins infra team to migrate the UC to new system and is **only** about copying data to remote systems, and **not** about the index generation, I'll proceed to merge:

- It has been reviewed by at least 2 other persons than me
- We tested it extensively
- The credentials have just been updated with latest Azure credentials

I'll watch the next update_center2 builds: if it fails, I'll either:
- hotfix this change and report here the hotfix PRs
- rollback the PR and analyse required changes.

**Comment by dduportal** [comment]
Backed up the `www` data to ensure a quick comeback if it goes wrong (in `/root/updates.jio-www-data-20240708-15h30-UTC.tgz`, integrity and content verified)

**Comment by dduportal** [comment]
Update: it went successfully and published the `gitea- checks` plugin with success:

https://plugins.jenkins.io/gitea-checks/releases/#version_470.v1d99442828cc

Capture d’écran 2024-07-08 à 17 39 34


The initial build time has been a bit long (~ 6min): currently checking next builds

**Comment by dduportal** [comment]
Next build was 3 min. Let's roll!

**Comment by dduportal** [comment]
Opened https://github.com/jenkins-infra/crawler/pull/148

**Pull request merged**